src/app/CommandSender.cpp

```diff
@@ -61,10 +62,22 @@ CHIP_ERROR CommandSender::SendCommandRequest(NodeId aNodeId, Transport::AdminId
     VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
     mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
 
-    err = mpExchangeCtx->SendMessage(
-        Protocols::InteractionModel::MsgType::InvokeCommandRequest, std::move(mCommandMessageBuf),
-        Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse).Set(Messaging::SendMessageFlags::kNoAutoRequestAck));
-    SuccessOrExit(err);
+    event.Type                              = DeviceLayer::DeviceEventType::kInteractionModelCommandRequest;
+    event.InteractionModelEvent.ExchangeCtx = mpExchangeCtx;
+
+    // ChipDeviceEvent can only contains the structures with trivial copy assignment operator, so we can't use PacketBufferHandle
+    // and SendFlags directly within ChipDeviceEvent.
+    event.InteractionModelEvent.Payload   = std::move(mCommandMessageBuf).UnsafeRelease();
+    event.InteractionModelEvent.SendFlags = Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)
+                                                .Set(Messaging::SendMessageFlags::kNoAutoRequestAck)
+                                                .Raw();
+
+    // Currently, Interaction Model protocol is implemented in app thread instead of CHIP thread, we need to lock CHIP stack to
+    // prevent it from operating before we post an event to the event queue associated with the CHIP thread.
+    DeviceLayer::PlatformMgr().LockChipStack();
+    DeviceLayer::PlatformMgr().PostEvent(&event);
```

Comment:
Why is the stack taking on the responsibility of providing 'event-ified' versions of these API calls? This is not a scalable pattern to have to add and maintain this extra boiler plate code, and arguably, it forces a very particular approach of how synchronization is to be achieved.

The application that is embedding the SDK should either be a) grabbing the CHIP lock and making these calls or b) creating 'posted' versions of these calls in their application as they see fit.